protoarray/src/test/java/tech/pegasys/teku/protoarray/AbstractBlockMetadataStoreTest.java

```diff
@@ -0,0 +1,260 @@
+    verifyHashesInChain(
+        store,
+        chainBuilder,
+        chainBuilder.getLatestBlockAndState().getRoot(),
+        chainBuilder.streamBlocksAndStates());
+
+    // And check we can start from part way along the chain
+    verifyHashesInChain(
+        store,
+        chainBuilder,
+        chainBuilder.getBlockAtSlot(6).getRoot(),
+        chainBuilder.streamBlocksAndStates(0, 6));
+  }
+
+  @Test
+  void processHashesInChainWhile_shouldStopProcessingWhenProcessorReturnsFalse() {
+    chainBuilder.generateBlocksUpToSlot(10);
+    final BlockMetadataStore store = createBlockMetadataStore(chainBuilder);
+
+    // And check we can start from part way along the chain
+    final Bytes32 headRoot = chainBuilder.getBlockAtSlot(6).getRoot();
+    final Stream<SignedBlockAndState> expectedBlocks = chainBuilder.streamBlocksAndStates(3, 6);
+
+    final Set<Bytes32> seenBlocks = new HashSet<>();
+
+    final AtomicReference<Bytes32> expectedBlock = new AtomicReference<>(headRoot);
+    store.processHashesInChainWhile(
+        headRoot,
+        (childRoot, slot, parentRoot) -> {
+          assertThat(seenBlocks).doesNotContain(childRoot);
+          seenBlocks.add(childRoot);
+
+          final SignedBeaconBlock block = chainBuilder.getBlock(expectedBlock.get()).orElseThrow();
+          assertThat(childRoot).isEqualTo(block.getRoot());
+          assertThat(slot).isEqualTo(block.getSlot());
+          assertThat(parentRoot).isEqualTo(block.getParentRoot());
+
+          expectedBlock.set(block.getParentRoot());
+          return !slot.equals(UInt64.valueOf(3));
+        });
+    // Check all fork blocks were seen
+    final Set<Bytes32> expectedSeenRoots =
+        expectedBlocks.map(StateAndBlockSummary::getRoot).collect(toSet());
+
+    assertThat(seenBlocks).containsExactlyInAnyOrderElementsOf(expectedSeenRoots);
+  }
+
+  private void verifyHashesInChain(
+      final BlockMetadataStore store,
+      final ChainBuilder chain,
+      final Bytes32 headRoot,
+      final Stream<SignedBlockAndState> expectedBlocks) {
+    final Set<Bytes32> seenBlocks = new HashSet<>();
+
+    final AtomicReference<Bytes32> expectedBlock = new AtomicReference<>(headRoot);
+    store.processHashesInChain(
+        headRoot,
+        (childRoot, slot, parentRoot) -> {
+          assertThat(seenBlocks).doesNotContain(childRoot);
+          seenBlocks.add(childRoot);
+
+          final SignedBeaconBlock block = chain.getBlock(expectedBlock.get()).orElseThrow();
+          assertThat(childRoot).isEqualTo(block.getRoot());
+          assertThat(slot).isEqualTo(block.getSlot());
+          assertThat(parentRoot).isEqualTo(block.getParentRoot());
+
+          expectedBlock.set(block.getParentRoot());
+        });
+    // Check all fork blocks were seen
+    final Set<Bytes32> expectedSeenRoots =
+        expectedBlocks.map(StateAndBlockSummary::getRoot).collect(toSet());
+
+    assertThat(seenBlocks).containsExactlyInAnyOrderElementsOf(expectedSeenRoots);
```

Comment:
Shouldn't we check that these were processed in order?

Reply:
We're tracking that we go from child to parent on each step which also checks order.